### PR TITLE
Support CentOS 7

### PR DESCRIPTION
**Pull request by bjoernhaeuser**
Adapt module to work better with CentOS7/RedHat 7

Also add configurable jmx/log4j/gc options to broker.

### CODE REVIEW

File: manifests/mirror/service.pp [hunk 1]
```diff
@@ -27,6 +27,15 @@
     content => template('kafka/mirror.init.erb'),
   }
 
+  if $::osfamily == 'RedHat' and $::operatingsystemmajrelease == '7' {
+    exec { 'systemctl daemon-reload # for kafka-mirror':
```
Comment (bastelfreak) [inline manifests/mirror/service.pp]: defining this exec multiple times isn't a good idea

**Comment by bastelfreak** [comment]
reloading systemd is always painful. I think there is a module on the forge that provides the exec with an "unless defined" wrapped around?m aybe we could use that instead?

File: manifests/mirror/service.pp (hunk 1 above)
Comment (bjoernhaeuser) [inline manifests/mirror/service.pp]: Its only defined once per mirror/producer/broker service.

**Comment by bjoernhaeuser** [comment]
Do you have a hint of the module? I looked into other voxpupuli modules and they also define the same exec.

**Comment by igalic** [comment]
must be one of these 11… https://forge.puppetlabs.com/modules?utf-8=%E2%9C%93&sort=rank&q=systemd

**Comment by jyaworski** [comment]
@igalic if the service file changes, you have to issue a daemon-reload.

**Comment by juniorsysadmin** [comment]
This one: https://forge.puppetlabs.com/camptocamp/systemd - but the namespace could have been named a little better

**Comment by bjoernhaeuser** [comment]
I added the camptocamp/systemd module. Tests are running fine, now trying the acceptance tests :)

File: manifests/broker/service.pp [hunk 2]
```diff
@@ -23,6 +26,12 @@
       content => template('kafka/init.erb'),
     }
 
+    if $::osfamily == 'RedHat' and $::operatingsystemmajrelease == '7' {
```
Comment (jyaworski) [inline manifests/broker/service.pp]: This would also be valid on Debian 8 and later Ubuntu. Can you use <code>$::serviceprovider<code> here?

Comment (bjoernhaeuser) [inline manifests/broker/service.pp]: Where is this fact defined? Cannot find it :-(

Comment (bjoernhaeuser) [inline manifests/broker/service.pp]: (And I actually dont like hardcoding these values here)

File: manifests/params.pp [hunk 3]
```diff
@@ -312,7 +317,6 @@
     'broker-list'                => '',
     'compression-codec'          => '',
     'key-serializer'             => 'kafka.serializer.DefaultEncoder',
-    'line-reader'                => 'kafka.tools.ConsoleProducer$LineMessageReader',
```
Comment (bjoernhaeuser) [inline manifests/params.pp]: This breaks the console producer in 0.9 :/

File: manifests/producer.pp [hunk 4]
```diff
@@ -46,6 +46,7 @@
 # }
 #
 class kafka::producer (
+  $input,
```
Comment (bjoernhaeuser) [inline manifests/producer.pp]: The console producer makes only sense if there is the possibility to read from something

File: manifests/broker/service.pp (hunk 2 above)
Comment (jyaworski) [inline manifests/broker/service.pp]: It's in stdlib:

https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/lib/facter/service_provider.rb

Comment (bjoernhaeuser) [inline manifests/broker/service.pp]: in Debian 8 the service provider is debian m(

Comment (bjoernhaeuser) [inline manifests/broker/service.pp]: Okay, only for puppet 3 :)

**Comment by bjoernhaeuser** [comment]
Using stdlib `$::service_provider` fact now

**Comment by bastelfreak** [comment]
Could you also replace the legacy sysv init script by a systemd service file? I'm not sure if upstream ships one, but several distributions already do, for example [Arch Linux](https://aur.archlinux.org/cgit/aur.git/tree/systemd_kafka.service?h=kafka).

**Comment by bjoernhaeuser** [comment]
@bastelfreak 

I added unit files for the broker-service. Could you please take a look and tell me if thats the right approach? If yes I would continue and provide unit files for the other services also.

File: manifests/producer/service.pp [hunk 5]
```diff
@@ -31,6 +32,12 @@
     content => template('kafka/producer.init.erb'),
   }
 
+  if $::service_provider == 'systemd' {
+    include ::systemd
+
+    File['/etc/init.d/kafka-producer'] ~> Exec['systemctl-daemon-reload'] -> Service['kafka-producer']
```
Comment (bastelfreak) [inline manifests/producer/service.pp]: we need to take care here about the unit file and set the init file to absent.

File: spec/classes/producer_spec.rb [hunk 6]
```diff
@@ -45,4 +47,40 @@
...
+
+    describe 'kafka::producer::config' do
+      context 'defaults' do
+        it { is_expected.to contain_file('/opt/kafka/config/producer.properties') }
+      end
+    end
+
+    describe 'kafka::producer::service' do
+      context 'defaults' do
+        it { is_expected.to contain_file('/etc/init.d/kafka-producer').that_notifies('Exec[systemctl-daemon-reload]') }
```
Comment (bastelfreak) [inline spec/classes/producer_spec.rb]: same as above, adjust to unit file path.

File: templates/broker.unit.erb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=Apache Kafka server (broker)
+Documentation=http://kafka.apache.org/documentation.html
+Requires=network.target remote-fs.target
+After=network.target remote-fs.target
```
Comment (bastelfreak) [inline templates/broker.unit.erb]: remote-fs.target isn't needed here. Does it require zookeeper?

Comment (bjoernhaeuser) [inline templates/broker.unit.erb]: Removed the remote-fs. Zookeeper is needed, but it normally does not run on the same machine.

**Comment by bjoernhaeuser** [comment]
Added all unit files \o/

File: manifests/broker/service.pp [hunk 8]
```diff
@@ -17,10 +20,27 @@
...
-    file { '/etc/init.d/kafka':
-      ensure  => present,
-      mode    => '0755',
-      content => template('kafka/init.erb'),
+    if $::service_provider == 'systemd' {
+      include ::systemd
+
+      file { '/usr/lib/systemd/system/kafka.service':
+        ensure  => present,
+        mode    => '0755',
```
Comment (bastelfreak) [inline manifests/broker/service.pp]: any reason for that mode? default for unit files is 0644.

File: manifests/producer/service.pp [hunk 9]
```diff
@@ -25,10 +26,15 @@
     fail('[Producer] You need to specify a value for topic')
   }
 
-  file { '/etc/init.d/kafka-producer':
-    ensure  => present,
-    mode    => '0755',
-    content => template('kafka/producer.init.erb'),
+  if $::service_provider == 'systemd' {
+    fail('Console Producer is not supported on systemd, because the stdin of the process cannot be redirected')
```
Comment (bastelfreak) [inline manifests/producer/service.pp]: I'm not exactly sure about the root of the issue. I think that should be possible. Can we discuss that in IRC?

File: templates/mirror.unit.erb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+Description=Apache Kafka server (mirror)
+Documentation=http://kafka.apache.org/documentation.html
+Requires=network.target
+After=network.target
+
+[Service]
+Type=simple
+User=kafka
+Group=kafka
+SyslogIdentifier=kafka-consumer
```
Comment (bastelfreak) [inline templates/mirror.unit.erb]: same identifier as consumer.unit.erb, is this on purpose?

File: manifests/producer/service.pp (hunk 9 above)
Comment (bastelfreak) [inline manifests/producer/service.pp]: will keep it for now. handling the producer with systemd is a bit dirty and in general rarely used. We can still implement this if it is requested

File: metadata.json [hunk 11]
```diff
@@ -18,11 +18,15 @@
     },
     {
       "name": "puppetlabs/stdlib",
-      "version_requirement": ">= 3.0.0"
+      "version_requirement": ">= 4.10.0"
```
Comment (bastelfreak) [inline metadata.json]: you've got 4.11 in fixtures. Can you bump that version here too?

**Comment by dhoppe** [comment]
@bjoernhaeuser Thank you very much for this great pull request.

**Pull request merged**